### PR TITLE
Enh/getcachednamespaces

### PR DESCRIPTION
**Pull request by oruebel**
## Motivation

#1431

Relate to https://github.com/dandi/dandi-cli/issues/917

This PR extracts the code to determine the cached extension namespace to use for validation from the main function of the command line validator to a separate function so that the code can be used outside to validate files against cached namespaces. 

## Checklist

- [x] Did you update CHANGELOG.md with your changes?
- [X] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [X] Have you ensured the PR clearly describes the problem and the solution?
- [X] Is your contribution compliant with our coding style? This can be checked running `flake8` from the source directory.
- [X] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/NeurodataWithoutBorders/pynwb/pulls) for the same change?
- [X] Have you included the relevant issue number using "Fix #XXX" notation where XXX is the issue number? By including "Fix #XXX" you allow GitHub to close issue #XXX when the PR is merged.

### CODE REVIEW

**Comment by oruebel** [comment]
@rly this issue would be good to address in the next minor release

**Comment by rly** [comment]
@oruebel what do you think about also extracting most of the code in `__main__` in `src/pynwb/validate.py` to a function that takes as inputs the arguments passed into the CLI function. Then a user can basically call the CLI script using Python without using the shell.

**Comment by oruebel** [comment]
> what do you think about also extracting most of the code in `__main__` in `src/pynwb/validate.py` to a function

that makes sense

**Comment by bendichter** [comment]
We agreed to incorporate this feature into `validate`

**Comment by yarikoptic** [comment]
would this PR be merged/released some time soon?

**Comment by CodyCBakerPhD** [comment]
@yarikoptic I'm close to finishing something very similar that also integrates with the `dandi.pynwb_utils.validate`, look for that in next day or so

File: src/pynwb/validate.py [hunk 1]
```diff
@@ -26,6 +26,52 @@ def _validate_helper(**kwargs):
...
+    """
+    catalog = NamespaceCatalog(NWBGroupSpec, NWBDatasetSpec, NWBNamespace)
+    ns_deps = NWBHDF5IO.load_namespaces(catalog, path)
+    # determine which namespaces are the most specific (i.e. extensions) and validate against those
+    s = set(ns_deps.keys())
+    for k in ns_deps:
+        s -= ns_deps[k].keys()
+    # TODO remove this workaround for issue https://github.com/NeurodataWithoutBorders/pynwb/issues/1357
+    if 'hdmf-experimental' in s:
+        s.remove('hdmf-experimental')  # remove validation of hdmf-experimental for now
```
Comment (jwodder) [inline src/pynwb/validate.py]:
```suggestion
    s.discard('hdmf-experimental')  # remove validation of hdmf-experimental for now
```

File: src/pynwb/validate.py [hunk 2]
```diff
@@ -26,6 +26,52 @@ def _validate_helper(**kwargs):
...
+    namespaces = list(sorted(s))
```
Comment (jwodder) [inline src/pynwb/validate.py]:
```suggestion
    namespaces = sorted(s)
```

`sorted()` already returns a list.

**Review by jwodder — COMMENTED** [review]
Pointed out some minor coding improvements.

**Comment by rly** [comment]
> Pointed out some minor coding improvements.

Thanks, @jwodder !

**Comment by yarikoptic** [comment]
this one is open since July -- any plans to finalize/release it ?

**Comment by oruebel** [comment]
> this one is open since July -- any plans to finalize/release it ?

@yarikoptic thanks for following up. I think this one got put on hold while #1494 was being worked out and I think that PR may potentially already address that issue. I'll bring it up during our developer meeting tomorrow to see whether this PR should be merged or whether it should be closed in favor of #1494

**Comment by CodyCBakerPhD** [comment]
@yarikoptic This has indeed already been merged in https://github.com/NeurodataWithoutBorders/pynwb/pull/1511, this can be closed here and I'll get a PR up on the dandi-cli to adjust the API calls to have this functionality

**Comment by oruebel** [comment]
Closing this PR since this has been addressed in https://github.com/NeurodataWithoutBorders/pynwb/pull/1511

**Pull request closed without merging**